src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ChannelFactoryBase.cs

```diff
@@ -186,5 +191,25 @@ protected override void OnEndClose(IAsyncResult result)
         {
             ChainedCloseAsyncResult.End(result);
         }
+
+        private async Task OnCloseAsyncInternal(TimeSpan timeout)
+        {
+            IChannel[] currentChannels = _channels.ToArray();
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+            foreach (IChannel channel in currentChannels)
+            {
+                IAsyncCommunicationObject iAsyncChannel = channel as IAsyncCommunicationObject;
+                if (iAsyncChannel != null)
+                {
+                    await iAsyncChannel.CloseAsync(timeout);
+                }
+                else
+                {
+                    channel.Close(timeoutHelper.RemainingTime());
+                }
+            }
```

Comment:
I think Close() is capable of throwing. Is there something we need to do to ensure handling of that case, if you're chaining a bunch of channel closes?